File: Tarea10_HeapSort_08-10-25/LEM_HeapSort.cs
using System;

class HeapSort
{
    // Función para ajustar el subárbol con raíz en el índice i
    static void Heapify(int[] arr, int n, int i)
    {
        int largest = i;        // Suponemos que la raíz es el mayor
        int left = 2 * i + 1;   // Hijo izquierdo
        int right = 2 * i + 2;  // Hijo derecho

        // Si el hijo izquierdo existe y es mayor que la raíz
        if (left < n && arr[left] > arr[largest])
            largest = left;

        // Si el hijo derecho existe y es mayor que el mayor actual
        if (right < n && arr[right] > arr[largest])
            largest = right;

        // Si el mayor no es la raíz, intercambiamos y seguimos ajustando recursivamente
        if (largest != i)
        {
            int temp = arr[i];
            arr[i] = arr[largest];
            arr[largest] = temp;

            // Llamada recursiva
            Heapify(arr, n, largest);
        }
    }

    // Función principal de heapsort
    static void HeapSortAlg(int[] arr)
    {
        int n = arr.Length;

        // Construir heap máximo
        for (int i = n / 2 - 1; i >= 0; i--)
            Heapify(arr, n, i);

        // Extraer elementos uno por uno
        for (int i = n - 1; i > 0; i--)
        {
            // Mover la raíz (máximo) al final
            int temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;

            // Llamar heapify en el heap reducido
            Heapify(arr, i, 0);
        }
    }

    static void Main()
    {
        int[] arr = { 15, 3, 66, 6, 44, 50 };
        int n = arr.Length;

        Console.Write("Antes de ordenar los elementos del arreglo: ");
        for (int i = 0; i < n; i++)
            Console.Write(arr[i] + " ");
        Console.WriteLine();

        HeapSortAlg(arr);

        Console.Write("\nDespués de ordenar el arreglo: ");
        for (int i = 0; i < n; i++)
            Console.Write(arr[i] + " ");
        Console.WriteLine();
    }
}

File: Tarea10_HeapSort_08-10-25/LEM_HeapSort.java
public class LEM_HeapSort {

    static void heapify(int arr[], int n, int i) {
        int largest = i;       
        int left = 2 * i + 1;   
        int right = 2 * i + 2;  

        if (left < n && arr[left] > arr[largest])
            largest = left;

        if (right < n && arr[right] > arr[largest])
            largest = right;

        if (largest != i) {
            int temp = arr[i];
            arr[i] = arr[largest];
            arr[largest] = temp;

            heapify(arr, n, largest);
        }
    }

    static void heapsort(int arr[]) {
        int n = arr.length;

        for (int i = n / 2 - 1; i >= 0; i--)
            heapify(arr, n, i);

        for (int i = n - 1; i > 0; i--) {
            int temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;

            heapify(arr, i, 0);
        }
    }

    public static void main(String[] args) {
        int arr[] = {15, 3, 66, 6, 44, 50};
        int n = arr.length;

        System.out.print("Antes de ordenar los elementos del arreglo: ");
        for (int i = 0; i < n; i++)
            System.out.print(arr[i] + " ");
        System.out.println();

        heapsort(arr);

        System.out.print("\nDespués de ordenar el arreglo: ");
        for (int i = 0; i < n; i++)
            System.out.print(arr[i] + " ");
        System.out.println();
    }
}
